homeassistant/components/legrand_rflc/config_flow.py

```diff
@@ -0,0 +1,171 @@
+from typing import Any, Final
+
+import lc7001.aio
+import voluptuous
+
+from homeassistant import config_entries, data_entry_flow
+from homeassistant.components.dhcp import IP_ADDRESS
+from homeassistant.const import (
+    CONF_AUTHENTICATION,
+    CONF_HOST,
+    CONF_MAC,
+    CONF_PASSWORD,
+    CONF_PORT,
+)
+from homeassistant.helpers.typing import DiscoveryInfoType
+
+from .const import DOMAIN
+
+_LOGGER: Final = logging.getLogger(__name__)
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """ConfigFlow for Legrand RFLC integration."""
+
+    VERSION = 1
+
+    HOST: Final = lc7001.aio.Connector.HOST
+
+    ABORT_NO_DEVICES_FOUND: Final = "no_devices_found"
+    ABORT_REAUTH_SUCCESSFUL: Final = "reauth_successful"
+
+    ERROR_INVALID_HOST: Final = "invalid_host"
+    ERROR_INVALID_AUTH: Final = "invalid_auth"
+
+    # Automatic ssdp discovery does not help.
+
+    # Automatic zeroconf discovery does not help.
+    # Even though the device supports mDNS, it does not publish any services to discover.
+
+    # Automatic dhcp discovery can detect conversations (udp port bootpc or bootps)
+    # with network devices that present their hostname as 'Legrand LC7001'
+    # with a manifest.json entry of
+    #   "dhcp": [{"hostname": "legrand lc7001"}]
+    # This will happen each time a Legrand LC7001 controller (re)boots.
+    # Linux requires the hass process have an effective cap_net_raw capability (or be run as root)
+    # for this to work.
+    async def async_step_dhcp(
+        self, discovery_info: DiscoveryInfoType
+    ) -> data_entry_flow.FlowResult:
+        """Handle a flow initiated by dhcp discovery."""
+        # example discovery_info
+        # {'ip': '192.168.0.1', 'hostname': 'legrand lc7001', 'macaddress': '0026ec000000'}
+        try:
+            resolutions = await asyncio.get_event_loop().getaddrinfo(self.HOST, None)
+        except OSError as error:
+            _LOGGER.warning("OS getaddrinfo %s error %s", self.HOST, error)
+            return self.async_abort(reason=self.ABORT_NO_DEVICES_FOUND)
+        address = discovery_info[IP_ADDRESS]
+        if any(
+            resolution[4][0] == address
+            for resolution in resolutions
+            if resolution[0] == socket.AF_INET
+        ):
+            await self._async_handle_discovery_without_unique_id()
+            # wait for user interaction in the next step
+            return await self.async_step_user()
+        _LOGGER.warning("%s does not resolve to discovered %s", self.HOST, address)
+        return self.async_abort(reason=self.ABORT_NO_DEVICES_FOUND)
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> data_entry_flow.FlowResult:
+        """Handle a flow initiated by the user."""
+        errors = {}
+        host = self.HOST
+        if user_input is not None:
+            host = user_input[CONF_HOST]
+            await self.async_set_unique_id(host)  # already_in_progress?
```

Comment:
Yes, I was aware of this guidance but, in this case, using the host name makes sense to me:

I want to be able to replace a failed Legrand RFLC LC7001 hub with a new (identically configured) one without having to change anything in Home Assistant.
As long as the user entered host name resolves to the new hub, this will work.
Using something like the MAC address will not.
I don't see the harm in doing this.

Reply:
If it doesn't follow our requirements as in our docs we don't allow it, sorry.

Reply:
This just doesn't make sense to me. I should be able to change the tire on my car without having to reconfigure the whole car. I don't see any other way to do this but to allow the user to name the device whatever he wants and make sure that when it is replaced it assumes same name.

Surely that is what is done all over the internet when the location of a service has changed. Its name doesn't change, it just gets resolved to somewhere else where the same service is given. The user of the service doesn't have to do a thing.

Reply:
I am not sure how this back and forth is supposed to work.

One of the nice features of Home Assistant (compared to SmartThings, where I am now) is the capability for the its "services" to be mobile in that I can move them (HA) from one piece of hardware to another. If the new HA hardware assumes the same name as the old one then nothing breaks in a larger system that has HA as a component. I am trying to do the same thing here except the hardware is the lc7001 hub (a HA component). I have had an lc7001 hub break on me. As the legrand_rflc software is now, moving to the lc7001 services to a new like-configured hub is/was done in the exact same manner. If it is a good thing for Home Assistant as a component, why is it a bad for thing for a component in Home Assistant?

The Legrand Lighting Control App(s) supports only one lc7001 hub. It is located/discovered using an mDNS lookup on the host name LCM1.local. That is, Legrand uses the device's host name to uniquely identify it.

Reply:
We need to be able to trust our API for unique_id. If the hostname changes for a device and we've used the hostname as unique_id of a config entry we can no longer trust that this API does what it is supposed to do.

IDs that can change are generally not allowed as unique_id unless we've documented it.

Reply:
Then we can just document this?

I am not sure what you mean by "trust our API". There is a Legrand RFLC API but that should only be of importance between the LC7001 hub, the lc7001 library (supported by me) and the legrand_rflc integration (supported by me). This should not concern HA or any HA API. I intend to support any LC7001 (discovered by host name) on the LAN using the Legrand RFLC API just as the Legrand Lighting Control App does on smart phones.

The only other thing that I could possibly do is use the MAC address. That would come with all the negative effects I have already described. That would also not fix anything if the user-configured hostname changes. I cannot locate an lc7001 by MAC address. I need a hostname. The mDNS hostname cannot be changed by the user. I allow for a user entered hostname because mDNS resolution does not work well (without complicated user-intervention) in HA docker containers AND, as an HA bonus feature exclusive, legrand_rflc allows for multiple lc7001's (on remote LANs that cannot be resolved by mDNS and coincident on the same LAN as in my home for supporting a production SmartThings environment and a HA testing environment at the same time).

Reply:
With "trust our API" I mean the config entries API in Home Assistant. It has this unique_id feature and it relies on the definition we have made regarding unique_id. We have defined unique_id as a string that should both be unique and stable for the connected device or service in the config entry. That means that the unique id is not allowed to be changed for the device or service during the config entry existence.

Reply:
The user has complete control over the LC7001 hub using the Legrand Lighting Control app on their phone. They can control its inventory of lights and affect their identifiers (name and ordered index). A careful owner that has to replace a broken hub can do so and maintain the same inventory that is identified in the same way. Both new and old hubs will mDNS-name themselves the same thing (LCM1.local). By default, LCM1.local is the hostname used by legrand_rflc (and the app). So, after replacement, the light controlled by same the LC7001 hostname at index N will be the same as before. This is a **very** good thing. One does not have to completely rebuild what they had working in HA. Unless they are a **very** careful user, they won't even remember how to do this is (especially when such configuration is external to HA and accessed though an API).

I assert that the care/effort needed by a user to recover what they had working in HA is much less if they are allowed to uniquely identify the hub by name. I am happy to document this if that is all that is required.

HA boasts the same quality with the ability to move its configuration from broken hardware to new hardware when the careful user properly executes a backup and restore. People have been complaining about this in SmartThings forever. Hooray for HA! Please tell me why such a capability is good for the goose (HA) but not for the gander (legrand_rflc).

Reply:
From a higher level HA perspective: `unique_id` is used like a primary key. If it is not unique there will be data corruption. If it can be changed and those changes aren't propagated everywhere, there will be data corruption. In general hostnames and ip addresses are not "stable" enough to meet the criteria for something that needs to be as good as a primary key. You have an interesting use case, but it is probably not compelling enough to the maintainers on its own to weaken this abstraction (in their eyes, it would be less reliable), especially as it only works for one integration.

For the stuff I contributed I use serial numbers read from the device, or the public part of a keypair if it has one. This works well, though doesn't satisfy your requirement.

I can imagine a tool in the settings panel that let you replace a device with another device, with vary degrees of automation, so any integration can support your use case. I can even imagine such a tool supporting the hub type changing (e.g. in theory no reason why I can't replace hue with homekit_controller and map the device replacements via serial number, potentially preserving even device automations in the process). That could be a really interesting project for someone.

Reply:
> In general hostname and IP addresses are not "stable" enough ...
They can be designed purposefully to be stable. Think google.com.

Unless there are further questions, I think I have pled my case as best as I can.
Are you implying that I am fighting an uphill battle and am doomed to die atop that hill?
Please tell me.

If so, I will dejectedly surrender now but I will gladly join you in fighting a battle to make this a general feature for like integrations. If it makes sense for HA as a component in a larger system it makes sense for a component in the HA system.

Reply:
> Think google.com.

Sure they can, but remember we are talking about the general case right now. In the general case, HA is designed to work on home networks. On my home network my laptop routinely manages to get confused and rename itself `laptop-2.local` because it thinks there was a collision (there wasnt). This is with "High End" gear, and after I (as a careful user) only named one computer "laptop.local". While Legrand RFLC coudl be the perfect IoT device, immune to any adverse network events, perfect in the face of even the most unreliable home network, these rules are there to protect HA from corrupting its own data, so they can't be ignored willy nilly. Think referential integrity.

I'm not going to put words in peoples mouths, but as someone who has been on a HA contribution right about where you are now let me say that these days I accept this particular rule gleefully and without contest.

I'm not advocating for battles, I think what I describe is a general feature that could be added withou disturbing anybody else.

Reply:
"Battle" just fit my metaphor. I don't want to fight anyone. I want to help make HA better. I do not doubt that is everyone's goal.

Multiple LC7001's in a multicast domain (say, a LAN) will collide on the same mDNS name (LCM1.local). Which one their app chooses seems to be a toss-up (which multicast response comes in first?). I see the same non-deterministic behavior from my mDNS capable name resolution in my Linux development environment. Since I do testing in such an environment (a production LC7001 used with SmartThings and a development LC7001 used with HA on the same LAN), I have purposefully configured the name resolution in this environment to resolve this collision. My system is purposefully designed so that this is not a problem. That is, the reference (hostname) uniquely identifies the desired instance and the expected services provided there are stable.

Admittedly, my setup is unusual. I would expect most setups would have only one LC7001 hub (uniquely named LCM1.local) and if there ever was another it would because it was replacing the first. It will assume the same mDNS name because that is what Legrand has chosen to do. That is, Legrand assumes this use case too. Legrand might later choose otherwise but Legrand has the power to break things in all sorts of ways (BTW, I have just been notified that they are breaking their Home+ Control API).

> I think what I describe is a general feature that could be added without disturbing anybody else.

I think that is wonderful but it is not here now.
I have a solution for this integration now.

Anyway, it seems as if my arguments are only convincing to me.

Reply:
https://github.com/home-assistant/core/pull/53430/commits/aeed68df34f4850fd9f942fa68f8636b691e4842

I am now using the discovered MAC address of the hub for the config entry's unique_id.
I am also using it as a prefix for every related light entity.
Is that correct?

Unfortunately, I don't think that I am done ...

What should I do when I find that I am connected to a different hub (I find a different MAC address). Right now, my reauth step says "invalid host" and asks the user to try again. The async_setup_entry, though, performs no check at all. One (or perhaps both) of these behaviors is probably not correct.